File: CH7(Backtracking)/Knight.cpp
#include <bits/stdc++.h>
using namespace std;

int n, k;

bool visited[10][10];

int dr[] = {2, 2, -2, -2, 1, 1, -1, -1};
int dc[] = {1, -1, 1, -1, 2, -2, 2, -2};

void solve(int row, int col, int cnt) {

    if (cnt == k) {
        visited[row][col] = true;
        return;
    }

    for (int i = 0; i < 8; i++) {

        int nr = row + dr[i];
        int nc = col + dc[i];

        if (nr >= 0 && nr < n && nc >= 0 && nc < n) {
            solve(nr, nc, cnt + 1);
        }
    }
}

int main() {

    cin >> n >> k;

    solve(0, 0, 0);

    int ans = 0;

    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (visited[i][j])
                ans++;
        }
    }

    cout << ans;

    return 0;
}

File: CH7(Backtracking)/Knight.py
n, k = map(int, input().split())

visited = [[False] * n for _ in range(n)]

dr = [2, 2, -2, -2, 1, 1, -1, -1]
dc = [1, -1, 1, -1, 2, -2, 2, -2]


def solve(row, col, cnt):

    if cnt == k:
        visited[row][col] = True
        return

    for i in range(8):

        nr = row + dr[i]
        nc = col + dc[i]

        if 0 <= nr < n and 0 <= nc < n:
            solve(nr, nc, cnt + 1)


solve(0, 0, 0)

ans = 0

for i in range(n):
    for j in range(n):
        if visited[i][j]:
            ans += 1

print(ans)